system/jhtree/jhtree.cpp

```diff
@@ -2345,86 +2381,105 @@ void CNodeCache::getCacheInfo(ICacheInfoRecorder &cacheInfo)
     leafCache.reportEntries(cacheInfo);
 }
 
-CJHTreeNode *CNodeCache::getNode(INodeLoader *keyIndex, unsigned iD, offset_t pos, IContextLogger *ctx, bool isTLK)
+CJHTreeNode *CNodeCache::getNode(INodeLoader *keyIndex, unsigned iD, offset_t pos, NodeType type, IContextLogger *ctx, bool isTLK)
 {
     // MORE - could probably be improved - I think having the cache template separate is not helping us here
     // Also one cache per key would surely be faster, and could still use a global total
     if (!pos)
         return NULL;
+
+    CKeyIdAndPos key(iD, pos);
+    NodeType cacheType = isTLK ? NodeBranch : type;
+    // This is what the code did previously - I suspect both should be ignored, or use separate caches
+    if ((cacheType == NodeMeta) || (cacheType == NodeBloom))
+        cacheType = NodeBranch;
+
     { 
-        // It's a shame that we don't know the type before we read it. But probably not that big a deal
         CriticalBlock block(lock);
-        CKeyIdAndPos key(iD, pos);
+        //MORE: Can this code be removed from the critical section?  Probably not because of parallel loading.
+        //This does not seem to be enabled in any of the createKeyIndex() calls - so likely to always be empty
         if (preloadNodes)
         {
             CJHTreeNode *cacheNode = preloadCache.query(key);
             if (cacheNode)
             {
+                //Only branch nodes and TLK leaf nodes are added to the preloadCache
+                assertex(cacheType == NodeBranch);
```

Comment:
Note: This means I can move the code into the switch branch below + will before final pr.